### PR TITLE
Support upto 4D array as REST input payload.

### LINKED ISSUE #5: Support N-dimensional data arrays as input in the request data array
Currently, the rest-proxy expects the input data array to be 1-D. However the KServe Predict v2 REST API [allows for passing in data in a multi-dimensional format](https://github.com/kserve/kserve/blob/master/docs/predict-api/v2/required_api.md#parameters).

With gRPC, however, the data must be the flattened, one-dimensional, row-major order of the tensor elements.

Some logic in the proxy is needed to flatten an arbitrary multi-dimensional array.

Example:

The following works since the array is flattened and 1-D.
```shell
curl -X POST -k http://localhost:8080/v2/models/example-mnist-predictor/infer -d '{ "id": "foo", "inputs": [{ "name": "predict", "shape": [2, 64], "datatype": "FP32", "data": [0.0, 0.0, 1.0, 11.0, 14.0, 15.0, 3.0, 0.0, 0.0, 1.0, 13.0, 16.0, 12.0, 16.0, 8.0, 0.0, 0.0, 8.0, 16.0, 4.0, 6.0, 16.0, 5.0, 0.0, 0.0, 5.0, 15.0, 11.0, 13.0, 14.0, 0.0, 0.0, 0.0, 0.0, 2.0, 12.0, 16.0, 13.0, 0.0, 0.0, 0.0, 0.0, 0.0, 13.0, 16.0, 16.0, 6.0, 0.0, 0.0, 0.0, 0.0, 16.0, 16.0, 16.0, 7.0, 0.0, 0.0, 0.0, 0.0, 11.0, 13.0, 12.0, 1.0, 0.0, 0.0, 0.0, 1.0, 11.0, 14.0, 15.0, 3.0, 0.0, 0.0, 1.0, 13.0, 16.0, 12.0, 16.0, 8.0, 0.0, 0.0, 8.0, 16.0, 4.0, 6.0, 16.0, 5.0, 0.0, 0.0, 5.0, 15.0, 11.0, 13.0, 14.0, 0.0, 0.0, 0.0, 0.0, 2.0, 12.0, 16.0, 13.0, 0.0, 0.0, 0.0, 0.0, 0.0, 13.0, 16.0, 16.0, 6.0, 0.0, 0.0, 0.0, 0.0, 16.0, 16.0, 16.0, 7.0, 0.0, 0.0, 0.0, 0.0, 11.0, 13.0, 12.0, 1.0, 0.0]}]}'
```

This doesn't yet work, but should:
```shell
curl -X POST -k http://localhost:8080/v2/models/example-mnist-predictor/infer -d '{ "id": "foo", "inputs": [{ "name": "predict", "shape": [2, 64], "datatype": "FP32", "data": [[0.0, 0.0, 1.0, 11.0, 14.0, 15.0, 3.0, 0.0, 0.0, 1.0, 13.0, 16.0, 12.0, 16.0, 8.0, 0.0, 0.0, 8.0, 16.0, 4.0, 6.0, 16.0, 5.0, 0.0, 0.0, 5.0, 15.0, 11.0, 13.0, 14.0, 0.0, 0.0, 0.0, 0.0, 2.0, 12.0, 16.0, 13.0, 0.0, 0.0, 0.0, 0.0, 0.0, 13.0, 16.0, 16.0, 6.0, 0.0, 0.0, 0.0, 0.0, 16.0, 16.0, 16.0, 7.0, 0.0, 0.0, 0.0, 0.0, 11.0, 13.0, 12.0, 1.0, 0.0],[0.0, 0.0, 1.0, 11.0, 14.0, 15.0, 3.0, 0.0, 0.0, 1.0, 13.0, 16.0, 12.0, 16.0, 8.0, 0.0, 0.0, 8.0, 16.0, 4.0, 6.0, 16.0, 5.0, 0.0, 0.0, 5.0, 15.0, 11.0, 13.0, 14.0, 0.0, 0.0, 0.0, 0.0, 2.0, 12.0, 16.0, 13.0, 0.0, 0.0, 0.0, 0.0, 0.0, 13.0, 16.0, 16.0, 6.0, 0.0, 0.0, 0.0, 0.0, 16.0, 16.0, 16.0, 7.0, 0.0, 0.0, 0.0, 0.0, 11.0, 13.0, 12.0, 1.0, 0.0]]}]}'
```

### PR DESCRIPTION
**Pull request by ScrapCodes**
closes #5

### CODE REVIEW

**Comment by ScrapCodes** [comment]
@pvaneck Please take a look !

**Comment by pvaneck** [comment]
Thanks @ScrapCodes , I will test this out soon. Currently having issues with my workstation I need to resolve. I'm not entirely sure why the lint job is suddenly failing with that `Running error: context loading failed: no go files to analyze` error. That will need to be looked into.

@njhill Would be great to have your eyes on this, too. :)

**Comment by ScrapCodes** [comment]
@pvaneck and @njhill thanks for taking a look, I have sent another PR to fix the CI. #7

**Comment by njhill** [comment]
Thanks @ScrapCodes. I think there should be a way we can streamline this and support arbitrary dimensions without having to deal with them as separate cases. I updated some experiments I was making before and can push that to a different branch to illustrate.

W.r.t. to test cases, I think we want to just have a set of raw json request payload examples along with the expected protobuf equivalent, and a bunch of raw protobuf response payload examples along with the expected REST equivalent. Just testing more cases end-to-end would be more valuable and simpler imo for this than more granular tests.

**Comment by njhill** [comment]
@ScrapCodes @pvaneck here's my proposed approach: https://github.com/kserve/rest-proxy/compare/main...njhill:dimensions?expand=1. It does not yet handle string / BYTES values though - I think there's some ambiguity in the spec around how BYTES values should/can be represented within the data array(s) so needs a bit more investigation.

**Comment by ScrapCodes** [comment]
Thanks @njhill, tests are not exactly as you had asked. Please take a look!

For handling BYTES in json, are you considering base64 encoding?

**Comment by njhill** [comment]
Thanks @ScrapCodes the tests are going in the right direction but I would suggest the following:
- Start with full REST request payload rather than just individual tensors, and test using the `CustomJSONPb.NewDecoder` func:
```go
out = &gw.ModelInferRequest{}
err := CustomJsonPb{}.NewDecoder(...).Decode(out)
// verify contents of out
```
- No need to duplicate the verification logic in separate functions, just put the jsons in an array and have the test func iterate over that to test each
- No need to diff individual fields of the target protobuf struct, can just compare the entire struct
- We should also test the response path in a similar way, i.e. `CustomJSONPb.Marshal`

For BYTES as mentioned I think the spec is currently ambiguous (to me at least). For example if strings are used then how should the encoding be determined. Also it's not clear how the byte lengths should be encoded when flattening. One option is to say that they must all be the same length and add a dimension to the shape for this. If their lengths can vary however then something extra is needed, like a corresponding integer tensor to hold them.

Previously I looked into how [MLServer](https://github.com/SeldonIO/MLServer) handles these kinds of inputs for some clues but do not recall the details now. Maybe you could look into that?

**Comment by njhill** [comment]
@ScrapCodes @pvaneck any thoughts/update on this? We should aim to get it in for the release right?

**Comment by pvaneck** [comment]
@njhill Yep, this should definitely be in the next release. I believe @ScrapCodes is currently finalizing the tests, then we can hopefully get this merged in. I think BYTES can be handled in a separate PR. I remember looking at some of MLServer's BYTES [string tests](https://github.com/SeldonIO/MLServer/blob/565b3b2a075b82b25aa7727de492a9056e8ee59c/tests/codecs/test_string.py)  and [base64 tests](https://github.com/SeldonIO/MLServer/blob/a284e937585b8f3db1ff8eac99b0b2b58c078398/tests/codecs/test_base64.py) before, but yea, this needs some more investigation for how we should handle it here.

File: proxy/marshaler_test.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+			t.Error(err)
+		}
+		i := out.Inputs[0]
+		if d := cmp.Diff(i.Name, expectedInput.Name); d != "" {
+			t.Errorf("diff: %s", d)
+		}
+		if d := cmp.Diff(i.Shape, inputDataShapes[k]); d != "" {
+			t.Errorf("diff: %s", d)
+		}
+		if d := cmp.Diff(i.Datatype, expectedInput.Datatype); d != "" {
```
Comment (ScrapCodes) [inline proxy/marshaler_test.go:132]: Comparing the entire struct at once did not work for some reason or the other. I have tried cmp options as well, trying to skip some fields did not help either.

File: proxy/marshaler_test.go [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+			],
+			[
+				[0.0, 0.0, 2.0, 12.0, 16.0, 13.0, 0.0, 0.0, 0.0, 0.0, 0.0, 13.0, 16.0, 16.0, 6.0, 0.0],
+				[0.0, 0.0, 0.0, 16.0, 16.0, 16.0, 7.0, 0.0, 0.0, 0.0, 0.0, 11.0, 13.0, 12.0, 1.0, 0.0]
+			]
+		]
+	]
+`
+
+func generateProtoBufResponse() gw.ModelInferResponse {
```
Comment (njhill) [inline proxy/marshaler_test.go:97]:
```suggestion
func generateProtoBufResponse() *gw.ModelInferResponse {
```

File: proxy/marshaler_test.go [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+	expectedOutput := []*gw.ModelInferResponse_InferOutputTensor{{
+		Name:     "predict",
+		Datatype: "INT64",
+		Shape:    []int64{2},
+		Contents: &gw.InferTensorContents{
+			Int64Contents: []int64{8, 8},
+		},
+	}}
+
+	return gw.ModelInferResponse{
```
Comment (njhill) [inline proxy/marshaler_test.go:107]:
```suggestion
	return &gw.ModelInferResponse{
```
@ScrapCodes this is what the linter is complaining about

**Comment by ScrapCodes** [comment]
@njhill thanks for taking a look, hope this is close to what we expect. Code looks cleaner and should perform faster thanks to you.

**Review by njhill — APPROVED** [review]
Thanks @ScrapCodes, lgtm

**Comment by njhill** [comment]
Thanks @pvaneck, I opened an issue https://github.com/kserve/rest-proxy/issues/8 for this.

Also I noticed that the files in this repo don't have the apache license/copyright headers, do you know if these are needed?

**Review by pvaneck — APPROVED** [review]
Thanks, @ScrapCodes

**Comment by pvaneck** [comment]
@njhill Ah, yea, they should be added. I can do that real quick.

**Pull request merged**